### PR TITLE
Ability to specify OS_* variables for OpenStack configuration

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

When we convert the OpenStack cloud provider to run in an external
process, we should be able to use kubernetes Secrets capability to
inject the OS_* variables. This way we can specify the cloud
configuration as a configmap, specify secrets for the userid/password
information. The configmap can be mounted as a file. the secrets can
be made available as environment variables. the external controller
itself can run as a pod/daemonset.

For backward compat, we preload all the OS_* variables, if anything
is in the config file, then that overrides the environment variables.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Authentication information for OpenStack cloud provider can now be specified as environment variables
```

### CODE REVIEW

**Comment by dims** [comment]
fyi, i am starting to experiment in https://github.com/dims/openstack-cloud-controller-manager

**Review by FengyunPan — APPROVED** [review]

**Comment by FengyunPan** [comment]
@anguslees PTAL, thanks.

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -180,12 +181,60 @@ func (cfg Config) toAuth3Options() tokens3.AuthOptions {
...
+	return
+}
+
 func readConfig(config io.Reader) (Config, error) {
 	if config == nil {
 		return Config{}, fmt.Errorf("no OpenStack cloud provider config file given")
 	}
 
-	var cfg Config
+	cfg, _ := configFromEnv()
```
Comment (dixudx) [inline pkg/cloudprovider/providers/openstack/openstack.go]: Does this mean we still have to read the config from env, even a config file is specified.

And if the config is not specified, should not we read the config from environment?

Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: @dixudx yes, the idea is to pre-populate the data structure what is in environment variables AND then overwrite the data present in the config file.

So if the config is not specified, we still read from environment

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 2]
```diff
@@ -180,12 +181,60 @@ func (cfg Config) toAuth3Options() tokens3.AuthOptions {
...
+// This allows acceptance testing against an existing OpenStack
+// install, using the standard OS_* OpenStack client environment
+// variables.
+func configFromEnv() (cfg Config, ok bool) {
+	cfg.Global.AuthUrl = os.Getenv("OS_AUTH_URL")
+
+	cfg.Global.TenantId = os.Getenv("OS_TENANT_ID")
+	// Rax/nova _insists_ that we don't specify both tenant ID and name
+	if cfg.Global.TenantId == "" {
+		cfg.Global.TenantName = os.Getenv("OS_TENANT_NAME")
```
Comment (huzhengchuan) [inline pkg/cloudprovider/providers/openstack/openstack.go]: can delete the code, the following will set cfg.Global.TenantId and cfg.Global.TenantId

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 3]
```diff
@@ -180,12 +181,60 @@ func (cfg Config) toAuth3Options() tokens3.AuthOptions {
...
+	}
+
```
Comment (huzhengchuan) [inline pkg/cloudprovider/providers/openstack/openstack.go]: I think can delete this code(from L190-L194), the following will set cfg.Global.TenantId and cfg.Global.TenantName

Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Thanks Done!

**Comment by dims** [comment]
/lgtm

(re-applying lgtm from @FengyunPan - needed to rebase)

**Comment by dims** [comment]
@FengyunPan looks like i can't apply the lgtm lost in rebase. can you please do that?

**Comment by FengyunPan** [comment]
Yeah, the PR is ok, and I'd like to approve it.
Wish the external cloud provider work is great.
/lgtm
/approve no-issue

**Pull request merged**